src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectCapabilityResolver.cs

```diff
@@ -120,20 +123,31 @@ private bool ContainingProjectHasCapability(string capability, string documentFi
             return false;
         }
 
+        var isMatch = false;
         try
         {
-            return vsHierarchy.IsCapabilityMatch(capability);
+            isMatch = vsHierarchy.IsCapabilityMatch(capability);
+
+            if (vsHierarchy.GetProjectFilePath(_jtf) is { } projectFilePath)
+            {
+                _cachedCapabilities = _cachedCapabilities.SetItem((projectFilePath, capability), isMatch);
+            }
         }
         catch (NotSupportedException)
         {
             // IsCapabilityMatch throws a NotSupportedException if it can't create a
             // BooleanSymbolExpressionEvaluator COM object
-            return false;
         }
         catch (ObjectDisposedException)
         {
             // IsCapabilityMatch throws an ObjectDisposedException if the underlying hierarchy has been disposed
-            return false;
         }
+
+        return isMatch;
+    }
+
+    public bool TryGetCachedCapabilityMatch(string projectFilePath, string capability, out bool isMatch)
```

Comment:
This `TryGetCached...` method feels a bit oogy. Callers of this will be depending on deterministic threading behavior by assuming that they won't execute until the UI thread has finished calling `ContainingProjectHasCapability`. Consider adding an async method that returns synchronously if the match is cached or asynchronously switches to the UI thread to calls `ContainingProjectHasCapability` if the match isn't cached.

* `ValueTask<bool> ProjectHasCapabilityAsync(string projectFileName, string capability, CancellationToken cancellationToken)`

Reply:
The whole reason this is a "try" is because it could return false negatives, and callers have to be okay with that. I had a note about that in a doc comment, but must have lost it (originally this method was elsewhere). I don't think its worth adding an async method like that, because it would have no callers. If it's needed in future, it could be added then.